sys/net/application_layer/nanocoap/sock.c

```diff
@@ -46,29 +64,41 @@ ssize_t nanocoap_request(coap_pkt_t *pkt, sock_udp_ep_t *local, sock_udp_ep_t *r
         return res;
     }
 
+    res = -EAGAIN;
+
     /* TODO: timeout random between between ACK_TIMEOUT and (ACK_TIMEOUT *
      * ACK_RANDOM_FACTOR) */
     uint32_t timeout = COAP_ACK_TIMEOUT * US_PER_SEC;
-    unsigned tries_left = COAP_MAX_RETRANSMIT + 1;  /* add 1 for initial transmit */
-    while (tries_left) {
+    uint32_t deadline = deadline_from_interval(timeout);
 
-        res = sock_udp_send(&sock, buf, pdu_len, NULL);
-        if (res <= 0) {
-            DEBUG("nanocoap: error sending coap request, %d\n", (int)res);
-            break;
+    /* add 1 for initial transmit */
+    unsigned tries_left = COAP_MAX_RETRANSMIT + 1;
+
+    while (tries_left) {
+        if (res == -EAGAIN) {
+            res = sock_udp_send(&sock, buf, pdu_len, NULL);
+            if (res <= 0) {
+                DEBUG("nanocoap: error sending coap request, %d\n", (int)res);
+                break;
+            }
         }
 
-        res = sock_udp_recv(&sock, buf, len, timeout, NULL);
+        res = sock_udp_recv(&sock, buf, len, deadline_left(deadline), NULL);
```

Comment:
I reckon this is to account for the time 'lost' by sending the UDP packet?

Reply:
This is to ensure that the total time listened is equal to the timeout, even when a recv call returned a packet with incorrect message ID.

Reply:
Better add a comment about it then.

Reply:
Added